### ISSUE TITLE
DNS propagation for domain with 69 subdomains

### ISSUE THREAD

**Issue by ktwalrus** [RoliSoft/LetsEncrypt-CloudFlare-Hook#4]
I'm experiencing a problem generating the certs for a domain with 69 subdomains. The CloudFlare hook sets the TXT record for each name challenge and waits for the DNS to propagate from the point of view of my server. When the letsencrypt.sh script tries to check the challenge, it occasionally fails and the script errors out. I have been unable to get all the way through the challenges to generate the certs and I now have hit some sort of limit on the number of pending challenges that I may have (so I guess I will have to wait a week to see if those failed challenges are removed and I can try again).

Looking at the letsencrypt.sh script, there is an option to bundle all the challenges into one call to hook.sh using the HOOK_CHAIN variable.

I think hook.sh should support the HOOK_CHAIN="yes" option to make the CF TXT records all at once, wait for them to propagate (checking that all the TXT records have propagated after all TXT records have been deployed) and then support an option to wait a couple more minutes before returning to letsencrypt.sh (increasing the chance that the LE servers are seeing the TXT records from their resolver).

This would greatly decrease the time it currently takes to wait on 70 TXT records to propagate since now, the hook.sh applies the records one at a time and waits on each record to propagate before proceeding to the next one. And, maybe avoid the problem of intermittent failure.

**Comment by RoliSoft** [comment]
I will look into supporting `HOOK_CHAIN="yes"` for such use-cases.

> increasing the chance that the LE servers are seeing the TXT records from their resolver

I don't think a timeout option is required, since LE is querying the origin nameservers directly at CloudFlare, not through a caching resolver. Therefore, if the script sees the change, the records have successfully propagated on CF's side and LE will also see them.

**Comment by RoliSoft** [comment]
I've added support for `HOOK_CHAIN` in 5c69214af6cb43e73244f0228ed42521ce414fc9.

Please test your setup on the staging server first, and don't hesitate to open a new issue if any problem arises.

**Comment by ktwalrus** [comment]
> I don't think a timeout option is required, since LE is querying the origin nameservers directly at CloudFlare, not through a caching resolver.

Well, I was experiencing intermittent failures with my 69 alt name cert that would happen every time somewhere into the list of challenges.

I had to wait a week before finally being successful after I changed the hook to only wait for the last challenge to be propagated, then wait an extra 2 minutes for good measure, and then have LE confirm all of the challenges one after another (still removing the TXT record at CF DNS after each confirmation).

If you implement HOOK_CHAIN="yes", then you can deploy all the challenges at CF without waiting for propagation, and then do a second pass over the challenges and confirm that all challenges have been propagated (which should be very fast since you only need to wait once and the other challenges should all be propagated already).

I suspect the intermittent problem of the hook seeing propagated records before LE see them is that either the two CF nameservers for my domain (eric.ns.cloudflare.com and erin.ns.cloudflare.com) are either not perfectly in sync or the CF origin nameservers are ANYCASTed and the server I'm connecting to is on the east coast while the server LE is checking is on the west coast. Actually, my computer was using 8.8.8.8 nameserver and this nameserver might be caching previous runs or be in sync with a different CF origin nameserver.

Maybe the real bug here is that hook.sh was checking only that the TXT record existed and not that the value of the TXT record matched the challenge. Since I was running hook.sh repeatedly retrying after failures, the old TXT records were cached by the 8.8.8.8 DNS nameserver dig was contacting (I have no idea what the TTL is for the TXT records), and hook.sh thought the new TXT record had propagated while it really hadn't.

Maybe you can add a check that not only is there a TXT record with the same name as the challenge being deployed, but that the TXT record has the same challenge text as the one the script deployed?

Also, you could lookup the nameserver for the domain:

dig ns +noall +answer $DOMAIN

and pick out the first NS listed and use dig @nameserver (e.g., dig @eric.ns.cloudflare.com ...) to lookup the TXT record so hook.sh isn't checking a cached resolver? This change with fully checking the TXT record value does have the same challenge test as it just deployed might fully fix this intermittent failure bug.

**Comment by RoliSoft** [comment]
> then you can deploy all the challenges at CF without waiting for propagation, and then do a second pass over the challenges and confirm that all challenges have been propagated

That's what the current implementation does, which I committed about an hour ago.

> Maybe you can add a check that not only is there a TXT record with the same name as the challenge being deployed, but that the TXT record has the same challenge text as the one the script deployed?

Yes, this was somewhat overlooked on my part. Since I don't have enough time to address it now, I'll reopen the issue and fix it in a few days.

> Also, you could lookup the nameserver for the domain

`dig +trace` does that by design, however that's also hitting the root nameservers. I'll see which implementation is more optimal and go with that.

**Comment by ktwalrus** [comment]
Okay. I'll wait to test what further changes you come up with to really make hook.sh bulletproof. I will be deploying hook.sh in a Docker container running periodically using cron. I need the script to be very reliable that I don't need to depend on someone noticing an intermittent failure... especially if the intermittent failures might lock me out from renewing a cert for a week (while the pending authorizations age off). The letsencrypt.sh script seems to renew certs when the old cert is less than 30 days from expiration and the automatic renewal process can probably tolerate a week or two delay, but this would not be optimal in production.

**Comment by the-solipsist** [comment]
Any update on this? I'm hitting errors of this kind randomly:

```
ERROR: Challenge is invalid! (returned: invalid) (result: {
  "type": "dns-01",
  "status": "invalid",
  "error": {
    "type": "urn:acme:error:unauthorized",
    "detail": "No TXT records found for DNS challenge",
    "status": 403
  },
  "uri": "https://acme-v01.api.letsencrypt.org/acme/challenge/[...]",
  "token": "[...]",
  "keyAuthorization": "[...]"
})
```

**Comment by RoliSoft** [comment]
Unfortunately I'm not working on this project anymore, as I had other issues with le.sh, and I ended up solving it by switching to https://github.com/Neilpang/acme.sh, which has built-in support for LetsEncrypt.
